modules/formulize/class/durationElement.php

```diff
@@ -0,0 +1,271 @@
+<?php
+require_once XOOPS_ROOT_PATH . "/modules/formulize/class/elements.php";
+require_once XOOPS_ROOT_PATH . "/modules/formulize/include/functions.php";
+
+class formulizeDurationElement extends formulizeElement
+{
+
+	function __construct()
+	{
+		$this->name = "Duration";
+		$this->hasData = true;
+		$this->needsDataType = false; // We'll force this to be stored as INT (minutes)
+		$this->overrideDataType = "int"; // Store as integer minutes
+		$this->adminCanMakeRequired = true;
+		$this->alwaysValidateInputs = true; // We'll always validate to ensure duration is within bounds
+		$this->canHaveMultipleValues = false;
+		$this->hasMultipleOptions = false;
+		parent::__construct();
+	}
+}
+
+class formulizeDurationElementHandler extends formulizeElementsHandler
+{
+
+	var $db;
+
+	function __construct($db)
+	{
+		$this->db = &$db;
+	}
+
+	// Conversion factors to minutes
+	private $timeUnits = array(
+		'days' => 1440,
+		'hours' => 60,
+		'minutes' => 1
+	);
+
+	function create()
+	{
+		return new formulizeDurationElement();
+	}
+
+	// Prepare admin UI data
+	function adminPrepare($element)
+	{
+		$dataToSendToTemplate = array();
+
+		if (is_object($element) && is_subclass_of($element, 'formulizeelement')) {
+			// Existing element
+			$ele_value = $element->getVar('ele_value');
+			$dataToSendToTemplate['ele_value'] = $ele_value;
+		} else {
+			// New element - set defaults
+			$ele_value = array(
+				'show_days' => 1,
+				'show_hours' => 1,
+				'show_minutes' => 1,
+				'min_minutes' => 0,
+				'max_minutes' => 0, // 0 means no maximum
+				'size' => 3 // width of each input box
+			);
+			$dataToSendToTemplate['ele_value'] = $ele_value;
+		}
+
+		return $dataToSendToTemplate;
+	}
+
+	// Save admin UI data
+	function adminSave($element, $ele_value)
+	{
+		if (is_object($element) && is_subclass_of($element, 'formulizeelement')) {
+			// Save which units to show
+			$ele_value['show_days'] = isset($_POST['show_days']) ? 1 : 0;
+			$ele_value['show_hours'] = isset($_POST['show_hours']) ? 1 : 0;
+			$ele_value['show_minutes'] = isset($_POST['show_minutes']) ? 1 : 0;
+
+			// Save min/max durations
+			$ele_value['min_minutes'] = intval($_POST['min_minutes']);
+			$ele_value['max_minutes'] = intval($_POST['max_minutes']);
+
+			// Save input box size
+			$ele_value['size'] = intval($_POST['size']);
+
+			$element->setVar('ele_value', $ele_value);
+
+			return true;
```

Comment:
So... the return true/false from adminSave is a bit nuanced, could perhaps be better explained in the comments... Basically, the issue comes down to the UI in the browser. The saving process is driven by asynchronous javascript requests. Page in browser doesn't change. But, if we're doing something funky here which would alter something that is in the admin UI in front of the user right now, then what they see on screen no longer matches the state in the database. If that has happened, then we should return true, because that will trigger a reload to the UI for the user, so the on screen and DB values match.

What this means in practice in this case, is that we would only return true, if the saved value of min_minutes or of max_minutes, what different from the $_POST-ed values. Same with size. Basically, if the intval resulted in a change to what the user sent, then we should update the page to match the currently saved values. If there was no change, no need to trigger a page load for the user, so we can return false.

Reply:
After reading through the code that calls the `adminSave` method I see what you mean. It definitely wasn't clear wtihout reading that what the return true/false values were meant to signify a value change that should result in a page reload. I had assumed that it was meant as a pass/fail boolean.